aii opennebula handle missing template

This error
```
[ERROR] Failed to execute plugin module's pxelinux Install method
[ERROR] Errors running plugin module pxelinux Install method: Can't call method "name" on an undefined value at /usr/lib/perl/AII/opennebula.pm line 650.
```
line 650 is the one with error in 
```perl
...
    if (@vmtpl) {
	$vmtemplate = $vmtpl[0];
        $main::this_app->info("Found VM template from ONE database: ", $vmtemplate->name);
    } else {
        $main::this_app->error("VM template is not available to be instantiated: ", $vmtemplate->name);
        return 0;
    }
...
```

--- a/aii-opennebula/src/main/pan/quattor/aii/opennebula/schema.pan
+++ b/aii-opennebula/src/main/pan/quattor/aii/opennebula/schema.pan
@@ -171,9 +171,10 @@ type opennebula_vmtemplate = {
     "diskcache" ? string with match(SELF, '^(default|none|writethrough|writeback|directsync|unsafe)$')
     "permissions" ? opennebula_permissions
     "pci" ? opennebula_vmtemplate_pci[]
-    @{labels is a string list to group the VMs under a given name and filter them 
+    @{labels is a list of strings to group the VMs under a given name and filter them 
     in the admin and cloud views. It is also possible to include in the list 
-    sub-labels using a common slash "Name, Name/SubName". This feature is available 
-    since OpenNebula 5.x, below this version the change does take any effect.}
+    sub-labels using a common slash: list("Name", "Name/SubName")
+    This feature is available since OpenNebula 5.x, below this version the change 
+    does not take effect.}
     "labels" ? string[]
 } = dict();

--- a/aii-opennebula/src/main/pan/quattor/aii/opennebula/schema.pan
+++ b/aii-opennebula/src/main/pan/quattor/aii/opennebula/schema.pan
@@ -171,4 +171,9 @@ type opennebula_vmtemplate = {
     "diskcache" ? string with match(SELF, '^(default|none|writethrough|writeback|directsync|unsafe)$')
     "permissions" ? opennebula_permissions
     "pci" ? opennebula_vmtemplate_pci[]
+    @{labels is a string list to group the VMs under a given name and filter them 
+    in the admin and cloud views. It is also possible to include in the list 
+    sub-labels using a common slash "Name, Name/SubName". This feature is available 
+    since OpenNebula 5.x, below this version the change does take any effect.}
+    "labels" ? string[]
 } = dict();

--- a/aii-opennebula/src/main/perl/opennebula.pm
+++ b/aii-opennebula/src/main/perl/opennebula.pm
@@ -637,7 +637,7 @@ sub install
         $vmtemplate = $vmtpl[0];
         $main::this_app->info("Found VM template from ONE database: ", $vmtemplate->name);
     } else {
-        $main::this_app->error("VM template is not available to be instantiated: ", $vmtemplate->name);
+        $main::this_app->error("VM template is not available to instantiate VM: $fqdn");
         return 0;
     }
 

--- a/aii-opennebula/src/main/perl/opennebula.pod
+++ b/aii-opennebula/src/main/perl/opennebula.pod
@@ -39,6 +39,8 @@ Modify your machine template to:
 
 =item * (Optional) Set /system/opennebula/graphics to export VM graphical display (VNC is used by default)
 
+=item * (Optional) Set /system/opennebula/labels list to include OpenNebula search labels
+
 =item * (Optional) Set /system/opennebula/pci list values to enable PCI Passthrough
 
 =over 6

--- a/aii-opennebula/src/main/resources/imagetemplate.tt
+++ b/aii-opennebula/src/main/resources/imagetemplate.tt
@@ -8,6 +8,9 @@ NAME = "[%- fqdn %]_[% pair.key %]"
 TARGET = "[% pair.key %]"
 SIZE = [% hardware.harddisks.${pair.key}.capacity %]
 DESCRIPTION = "QUATTOR image for [% fqdn %]: [% pair.key %]"
+[%- IF system.opennebula.labels.defined %]
+LABELS = "[% system.opennebula.labels.join(',') %]"
+[%- END %]
 QUATTOR = 1
 [% # If we have several disks we have to split vmtemplate in several files
    # Using DATASTORE line. THIS MUST BE THE LAST TT FILE LINE-%]

--- a/aii-opennebula/src/main/resources/tests/profiles/vm.pan
+++ b/aii-opennebula/src/main/resources/tests/profiles/vm.pan
@@ -164,3 +164,8 @@ prefix "/system/opennebula";
     "device", 0x0a0c,
     "class", 0x0403,
 ));
+
+"labels" = list (
+    "quattor",
+    "quattor/CE",
+);

--- a/aii-opennebula/src/main/resources/tests/regexps/aii_imagetemplate/image_template
+++ b/aii-opennebula/src/main/resources/tests/regexps/aii_imagetemplate/image_template
@@ -13,5 +13,6 @@ multiline
 ^TARGET\s?=\s?"vdb"$
 ^SIZE\s?=\s?10480$
 ^DESCRIPTION\s?=\s?"QUATTOR image for node630.cubone.os: vdb"$
+^LABELS\s?=\s?"quattor,quattor/CE"$
 ^QUATTOR\s?=\s?1$
 ^DATASTORE\s?=\s?"default"$

--- a/aii-opennebula/src/main/resources/tests/regexps/aii_imagetemplate/simple
+++ b/aii-opennebula/src/main/resources/tests/regexps/aii_imagetemplate/simple
@@ -9,5 +9,6 @@ multiline
 ^TARGET\s?=\s*".+"\s*$
 ^SIZE\s?=\s*\d+\s*$
 ^DESCRIPTION\s?=\s*".+"\s*$
+^LABELS\s?=\s?".+"$
 ^QUATTOR\s?=\s?\d?\s*$
 ^DATASTORE\s?=\s*".+"\s*$

--- a/aii-opennebula/src/main/resources/tests/regexps/aii_vmtemplate/simple
+++ b/aii-opennebula/src/main/resources/tests/regexps/aii_vmtemplate/simple
@@ -37,4 +37,5 @@ multiline
 ^\s*CLASS\s?=\s*".+",\s*$
 ^\s*DEVICE\s?=\s*".+",\s*$
 ^\s*VENDOR\s?=\s*".+"\s*$
+^LABELS\s?=\s*".+"$
 ^QUATTOR\s?=\s?\d+\s*$

--- a/aii-opennebula/src/main/resources/tests/regexps/aii_vmtemplate/simple
+++ b/aii-opennebula/src/main/resources/tests/regexps/aii_vmtemplate/simple
@@ -24,6 +24,7 @@ multiline
 ^\s*IMAGE_UNAME\s?=\s*".+"\s*$
 ^GRAPHICS\s?=\s?\[$
 ^\s*LISTEN\s?=\s*".+",\s*$
+^\s*RANDOM_PASSWD\s?=\s*".+",\s*$
 ^\s*TYPE\s?=\s*".+"\s*$
 ^MEMORY\s?=\s*"\d+"\s*$
 ^OS\s?=\s?\[$

--- a/aii-opennebula/src/main/resources/tests/regexps/aii_vmtemplate/vm_template
+++ b/aii-opennebula/src/main/resources/tests/regexps/aii_vmtemplate/vm_template
@@ -64,4 +64,5 @@ multiline
 ^\s{4}\<devices\>\<video\>\<model\s{1}type\=\'qxl\'\/\>\<\/video\>\<\/devices\>",$
 ^\s{4}TYPE\s?=\s?"kvm"$
 ^\]$
+^LABELS\s?=\s?"quattor,quattor/CE"$
 ^QUATTOR\s?=\s?1$

--- a/aii-opennebula/src/main/resources/tests/regexps/aii_vmtemplate/vm_template
+++ b/aii-opennebula/src/main/resources/tests/regexps/aii_vmtemplate/vm_template
@@ -52,6 +52,7 @@ multiline
 ^\]$
 ^GRAPHICS\s?=\s?\[$
 ^\s{4}LISTEN\s?=\s?"0.0.0.0",$
+^\s{4}RANDOM_PASSWD\s?=\s?"YES",$
 ^\s{4}TYPE\s?=\s?"SPICE"$
 ^\]$
 ^MEMORY\s?=\s?"4096"$

--- a/aii-opennebula/src/main/resources/vmtemplate.tt
+++ b/aii-opennebula/src/main/resources/vmtemplate.tt
@@ -54,6 +54,7 @@ IMAGE_UNAME = "oneadmin"
 GRAPHICS = [
 [%     FILTER indent -%]
 LISTEN = "0.0.0.0",
+RANDOM_PASSWD = "YES",
 TYPE = "[% system.opennebula.graphics %]"
 [%-    END %]
 ]

--- a/aii-opennebula/src/main/resources/vmtemplate.tt
+++ b/aii-opennebula/src/main/resources/vmtemplate.tt
@@ -85,4 +85,7 @@ PCI = [
 [%-         END %]
 [%-     END %]
 [%- END %]
+[%- IF system.opennebula.labels.defined %]
+LABELS = "[% system.opennebula.labels.join(',') %]"
+[%- END %]
 QUATTOR = 1

--- a/aii-opennebula/src/test/perl/rpcdata.pm
+++ b/aii-opennebula/src/test/perl/rpcdata.pm
@@ -596,6 +596,7 @@ DISK = [
 ]
 GRAPHICS = [
     LISTEN = "0.0.0.0",
+    RANDOM_PASSWD = "YES",
     TYPE = "SPICE"
 ]
 MEMORY = "4096"

--- a/aii-opennebula/src/test/perl/rpcdata.pm
+++ b/aii-opennebula/src/test/perl/rpcdata.pm
@@ -490,6 +490,7 @@ NAME = "node630.cubone.os_vda"
 TARGET = "vda"
 SIZE = 20480
 DESCRIPTION = "QUATTOR image for node630.cubone.os: vda"
+LABELS = "quattor,quattor/CE"
 QUATTOR = 1
 EOF
 $cmds{rpc_create_image}{params} = [$data, 102];
@@ -505,6 +506,7 @@ NAME = "node630.cubone.os_vdb"
 TARGET = "vdb"
 SIZE = 10480
 DESCRIPTION = "QUATTOR image for node630.cubone.os: vdb"
+LABELS = "quattor,quattor/CE"
 QUATTOR = 1
 EOF
 $cmds{rpc_create_image2}{params} = [$data, 102];
@@ -621,6 +623,7 @@ PCI = [
     DEVICE = "1616",
     VENDOR = "8086"
 ]
+LABELS = "quattor,quattor/CE"
 QUATTOR = 1
 EOF
 $cmds{rpc_create_template}{params} = [$data];

--- a/aii-opennebula/src/test/resources/vm.pan
+++ b/aii-opennebula/src/test/resources/vm.pan
@@ -144,3 +144,8 @@ prefix "/system/opennebula";
     "device", 0x0a0c,
     "class", 0x0403,
 ));
+
+"labels" = list (
+    "quattor",
+    "quattor/CE",
+);

